// functional-mini/dist/esm/component.js
import { functionalMiniElement } from './element.js';
import { getLifeCycleHooks } from './hooks.js';
import { alipayComponentEvents, commonComponentEvents } from './platform.js';
import { EElementType, ETargetPlatform } from './types.js';
export function alipayComponent(element, defaultProps) {
    return functionalMiniElement(element, '', EElementType.component, defaultProps, ETargetPlatform.alipay);
}
export function wechatComponent(element, defaultProps) {
    return functionalMiniElement(element, '', EElementType.component, defaultProps, ETargetPlatform.wechat);
}
export var useAttached = getLifeCycleHooks(commonComponentEvents.attached);
export var useReady = getLifeCycleHooks(commonComponentEvents.ready);
export var useMoved = getLifeCycleHooks(commonComponentEvents.moved);
export var useDidMount = getLifeCycleHooks(alipayComponentEvents.didMount, undefined, ETargetPlatform.alipay);
// onInit, deriveDataFromProps, didUnmount, didUpdate 是不开放 Hooks 的，请使用 React 生命周期方法
export * from './export-hooks.js';
export { useComponent } from './hooks.js';
